### PR TITLE
Tests(Cronjobs): Adding tests for MonitorLogs and MonitorHoneyPots.

### PR DESCRIPTION
**Pull request by amishhaa**
Fixes: #653
# Description

This PR adds tests for monitor_logs.py and monitor_honeypots.py.

## Related issues
#653
## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
# Checklist

- [x] I have read and understood the rules about [how to Contribute](https://intelowlproject.github.io/docs/GreedyBear/Contribute/) to this project.
- [x] The pull request is for the branch `develop`.
- [ ] I have added documentation of the new features.
- [x] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowlproject.github.io/docs/GreedyBear/Contribute/#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [x] I have added tests for the feature/bug I solved. All the tests (new and old ones) gave 0 errors.
- [ ] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check [CONTRIBUTE.md](https://github.com/intelowlproject/docs/blob/main/docs/GreedyBear/Contribute.md)).
- [ ] If the GUI has been modified:
    - [ ] I have a provided a screenshot of the result in the PR.
    - [ ] I have created new frontend tests for the new component or updated existing ones.
  
### Important Rules
- If you miss to compile the Checklist properly, your PR won't be reviewed by the maintainers.
- If your changes decrease the overall tests coverage (you will know after the Codecov CI job is done), you should add the required tests to fix the problem
- Everytime you make changes to the PR and you think the work is done, you should explicitly ask for a review. After being reviewed and received a "change request", you should explicitly ask for a review again once you have made the requested changes.

### CODE REVIEW

**Comment by amishhhaaaa** [comment]
Hey @regulartim ! thank you for pointing me in the right direction in the previous PR. I think this is in a nice shape for a review.

File: tests/greedybear/cronjobs/test_monitor_honeypots.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        # Run the cronjob
+        cronjob.execute()
+
+        self.assertEqual(mock_elastic_repo.has_honeypot_been_hit.call_count, 2)
+
+        info_calls = [call[0][0] for call in cronjob.log.info.call_args_list]
+        warning_calls = [call[0][0] for call in cronjob.log.warning.call_args_list]
+
+        self.assertTrue(any("logs available" in msg for msg in info_calls))
+        self.assertTrue(any("no logs available" in msg for msg in warning_calls))
```
Comment (regulartim) [inline tests/greedybear/cronjobs/test_monitor_honeypots.py]: Wait, you are testing the output of `cronjob.log` here, but that is a `MagicMock` right? So this will always succeed! Or am I on the wrong track here?

File: tests/greedybear/cronjobs/test_monitor_logs.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+        # Side effect for multiple calls
+        def stat_side_effect():
+            mock_stat_result = MagicMock()
+            if stat_side_effect.call_count == 0:
+                mock_stat_result.st_mtime = recent_time
+            else:
+                mock_stat_result.st_mtime = old_time
+            stat_side_effect.call_count += 1
+            return mock_stat_result
```
Comment (regulartim) [inline tests/greedybear/cronjobs/test_monitor_logs.py]: I don't understand what this does. Could you please explain?

File: tests/greedybear/cronjobs/test_monitor_logs.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+from greedybear.cronjobs.monitor_logs import MonitorLogs
+from greedybear.slack import send_message
+
+
+class MonitorLogsTestCase(TestCase):
+    @patch("greedybear.cronjobs.monitor_logs.Path.stat")
+    @patch("greedybear.cronjobs.monitor_logs.Path.exists")
+    @patch("greedybear.cronjobs.monitor_logs.send_message")
+    def test_logs_recent_activity(self, mock_send, mock_exists, mock_stat):
```
Comment (regulartim) [inline tests/greedybear/cronjobs/test_monitor_logs.py]: Could you also write a test, where the log was not populated and nothing was sent?

**Comment by regulartim** [comment]
First of all, thanks for your work. But I think there is a lot to be done here. I commented on some parts of the code, but also in the bigger picture your code is not ready in my opinion. Here are a few tips for you:
- one test case per class is not enough: test edge cases, happy paths and failing paths
- try to stay consistent with the structure of existing tests
- use more precise naming: is something is named "...IntegrationTest" it should be an integration test

File: tests/greedybear/cronjobs/test_monitor_logs.py (hunk 2 above)
Comment (amishhhaaaa) [reply tests/greedybear/cronjobs/test_monitor_logs.py]: Yes! So if you refer line https://github.com/intelowlproject/GreedyBear/blob/develop/greedybear/cronjobs/monitor_logs.py#L27 we are checking the logs for 4 different pots, so stat_result which contains the time is called four times, so this function returns recent time for the first call and old time for rest. Hence later we assert that the send_message method was only called once with greedy bear as it was the only pot with latest log time.......this can be decoupled into two test functions surely but as already iterated by you that we prefer to follow the test format present, i took reference from test https://github.com/intelowlproject/GreedyBear/blob/develop/tests/greedybear/cronjobs/test_firehol.py where multiple test cases were written in side effect.

File: tests/greedybear/cronjobs/test_monitor_logs.py (hunk 3 above)
Comment (amishhhaaaa) [reply tests/greedybear/cronjobs/test_monitor_logs.py]: yes! @regulartim would you prefer the general test design where we write multiple functions and each method is responsible for testing a specific behavior or would you prefer it to how it is currently and all the tests coupled in one single function using side effects?

File: tests/greedybear/cronjobs/test_monitor_honeypots.py (hunk 1 above)
Comment (amishhhaaaa) [reply tests/greedybear/cronjobs/test_monitor_honeypots.py]: I used MagicMock to make a mock object of cronjob.log so we dont actually log something, apart from that the behavior of it depends only on what is being logged by the upper methods that call it, so if your concern was did i mock and pass the log manually, so no that is actually not happening from my understanding. I feel this class can be improved a lot and the tests can be more clear with what they are doing. Thanks for the feedback!

File: tests/greedybear/cronjobs/test_monitor_logs.py (hunk 3 above)
Comment (regulartim) [reply tests/greedybear/cronjobs/test_monitor_logs.py]: I would prefer multiple functions and each method is responsible for testing a specific behavior.

File: tests/greedybear/cronjobs/test_monitor_logs.py (hunk 2 above)
Comment (regulartim) [reply tests/greedybear/cronjobs/test_monitor_logs.py]: I would prefer multiple functions and each method is responsible for testing a specific behavior.

File: tests/greedybear/cronjobs/test_monitor_honeypots.py (hunk 1 above)
Comment (regulartim) [reply tests/greedybear/cronjobs/test_monitor_honeypots.py]: Ah, ok. I think I understand. I think I was confused because I expected that it is tested which honeypot logged what. You verify that some info and warning were logged, but not that Log4pot got the info and Cowrie got the warning. Because this is what happens, right?

**Comment by amishhhaaaa** [comment]
hey @regulartim, i have made multiple test cases to reflect the specific behavior that is being tested, also improved the naming. let me know if any further clarification or changes are required. Thank you again for the through feedback!

File: tests/greedybear/cronjobs/test_monitor_honeypots.py (hunk 1 above)
Comment (amishhhaaaa) [reply tests/greedybear/cronjobs/test_monitor_honeypots.py]: @regulartim i think i understand, as we know the name of the honeypot we can actually log the name of it in the assertion, the latest commit actually asserts equal with the number directly, but logging the name is also doable, might make the tests brittle if we do it but also enhances the code from readability point of view because we are clear just by reading whats happening. let me know what you prefer i would assert it accordingly.

File: tests/greedybear/cronjobs/test_monitor_logs.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
+from datetime import datetime, timedelta
+from pathlib import Path
+from unittest import TestCase
+from unittest.mock import MagicMock, patch
+
+from greedybear.cronjobs.monitor_logs import MonitorLogs
+from greedybear.slack import send_message
```
Comment (regulartim) [inline tests/greedybear/cronjobs/test_monitor_logs.py]: Isn't this unused?

File: tests/greedybear/cronjobs/test_monitor_logs.py (hunk 2 above)
Comment (regulartim) [reply tests/greedybear/cronjobs/test_monitor_logs.py]: I am sorry, but I still find that function very confusing, although I do understand now why it is here and what it is doing. Would something like this be equivalent?
```python
mock_stat.side_effect = [
       MagicMock(st_mtime=recent_time),
       MagicMock(st_mtime=old_time),
       MagicMock(st_mtime=old_time),
       MagicMock(st_mtime=old_time),
   ]
```

I think this is more explicit and easier to read and understand. What do you think?

**Review by regulartim — CHANGES_REQUESTED** [review]
Good job! We're getting close! :)

File: tests/greedybear/cronjobs/test_monitor_logs.py (hunk 4 above)
Comment (amishhhaaaa) [reply tests/greedybear/cronjobs/test_monitor_logs.py]: yes removed, very surprised it wasnt yellow lined as unused import by my ide or the pre-commit checks, great catch!!

**Comment by amishhhaaaa** [comment]
@regulartim i have made the necessary changes :D, sorry for the delay

**Comment by amishhhaaaa** [comment]
@regulartim I went through the test file of firehol.py, i believe the tests in that can be decoupled and more edges can be tested also that file no longer strongly follow the format we have established in these test files like seperation of concerns, would you like if i make an issue highlighting some concerns, if you feel its worth working on i would like to take that up.

**Comment by regulartim** [comment]
> @regulartim I went through the test file of firehol.py, i believe the tests in that can be decoupled and more edges can be tested also that file no longer strongly follow the format we have established in these test files like seperation of concerns, would you like if i make an issue highlighting some concerns, if you feel its worth working on i would like to take that up.

Would you please open a separate issue on that?

**Comment by regulartim** [comment]
> @regulartim i have made the necessary changes :D, sorry for the delay

No problem! :)

**Comment by regulartim** [comment]
For future PRs: when I am asking questions in a review, I would really like to discuss the code with you. Sometimes I'm wrong or I am interested in your rationale.

**Comment by amishhhaaaa** [comment]
yes! @regulartim i will open a new issue shortly. yes i would take care in future PRs, thanks again for the thorough review!
